File: apps/kira-studio/internal/gitpreflight/operation_test.go
package gitpreflight_test

import (
	"testing"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/gitpreflight"
)

func strp(s string) *string { return &s }

func TestClassifyInProgress_NoState(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{}, nil)
	if op != nil {
		t.Fatalf("op = %+v, want nil", op)
	}
}

func TestClassifyInProgress_Merge(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{MergeHead: strp("abc123")}, nil)
	if op == nil || op.Kind != gitpreflight.InProgressMerge {
		t.Fatalf("op = %+v", op)
	}
	if !op.CanContinue || !op.CanAbort || op.CanSkip {
		t.Fatalf("op = %+v, want canContinue/canAbort true, canSkip false", op)
	}
	if op.OtherSha == nil || *op.OtherSha != "abc123" {
		t.Fatalf("otherSha = %v", op.OtherSha)
	}
}

func TestClassifyInProgress_CherryPick(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{CherryPickHead: strp("c1")}, nil)
	if op == nil || op.Kind != gitpreflight.InProgressCherryPick || !op.CanSkip {
		t.Fatalf("op = %+v, want cherryPick with canSkip", op)
	}
}

func TestClassifyInProgress_Revert(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{RevertHead: strp("r1")}, nil)
	if op == nil || op.Kind != gitpreflight.InProgressRevert || !op.CanSkip {
		t.Fatalf("op = %+v, want revert with canSkip", op)
	}
}

// TestClassifyInProgress_RebaseOffersContinueAndSkip_G26 inverts G5's original
// TestClassifyInProgress_Rebase, which asserted canContinue/canSkip both false under "§9's
// report-only posture". G26 D12 retires that posture: G26's own restack executor is the first
// thing in this app that can START a rebase, so a paused restack needs Continue (and Skip, for an
// already-applied commit) to have any exit besides a terminal. Renamed, not deleted, per plan §6
// step 3/D12's own instruction.
func TestClassifyInProgress_RebaseOffersContinueAndSkip_G26(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{
		RebaseMergeDir: true, RebaseHeadName: strp("refs/heads/side"), RebaseOnto: strp("onto-sha"),
	}, nil)
	if op == nil || op.Kind != gitpreflight.InProgressRebase {
		t.Fatalf("op = %+v", op)
	}
	if !op.CanContinue {
		t.Fatal("G26 D12: rebase must now offer Continue (retires §9's report-only posture)")
	}
	if !op.CanAbort {
		t.Fatal("rebase should offer Abort")
	}
	if !op.CanSkip {
		t.Fatal("G26 D12: rebase must now offer Skip (probe P10's own hint names it)")
	}
	if op.HeadName == nil || *op.HeadName != "refs/heads/side" {
		t.Fatalf("headName = %v", op.HeadName)
	}
}

// TestClassifyInProgress_RebaseShadowsSequencerFiles proves the precedence table's own shadowing
// rule: a rebase stopped on a conflict can ALSO leave sequencer-shaped files behind, but rebase
// must still win.
func TestClassifyInProgress_RebaseShadowsSequencerFiles(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{
		RebaseMergeDir: true, CherryPickHead: strp("would-be-cherry-pick"), SequencerDir: true,
	}, nil)
	if op == nil || op.Kind != gitpreflight.InProgressRebase {
		t.Fatalf("op = %+v, want rebase to shadow cherryPick", op)
	}
	if !op.IsSequence {
		t.Fatal("isSequence should still be true")
	}
}

func TestClassifyInProgress_Bisect(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{BisectLog: true}, nil)
	if op == nil || op.Kind != gitpreflight.InProgressBisect {
		t.Fatalf("op = %+v", op)
	}
	if op.CanContinue || op.CanSkip {
		t.Fatal("bisect offers neither continue nor skip")
	}
}

// TestClassifyInProgress_UnmergedOnlyWithNoStateFile proves the fallback case: unmerged paths with
// none of the six state files present (a resolved-then-reset state, or `checkout -m`).
func TestClassifyInProgress_UnmergedOnlyWithNoStateFile(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{}, []string{"a.txt", "b.txt"})
	if op == nil || op.Kind != gitpreflight.InProgressUnmergedOnly {
		t.Fatalf("op = %+v", op)
	}
	if op.CanContinue || op.CanAbort {
		t.Fatal("unmergedOnly offers neither continue nor abort")
	}
	if op.UnmergedCount != 2 {
		t.Fatalf("unmergedCount = %d, want 2", op.UnmergedCount)
	}
}

// TestClassifyInProgress_CanContinueIndependentOfUnmergedCount proves canContinue/canSkip are
// kind-level facts, never derived from conflictedPaths/unmergedCount — the banner's own enablement
// rule (canContinue && unmergedCount === 0) depends on the two staying independent.
func TestClassifyInProgress_CanContinueIndependentOfUnmergedCount(t *testing.T) {
	withConflicts := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{MergeHead: strp("m")}, []string{"x.txt"})
	withoutConflicts := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{MergeHead: strp("m")}, nil)
	if !withConflicts.CanContinue || !withoutConflicts.CanContinue {
		t.Fatalf("canContinue should be true regardless of unmergedCount: %+v / %+v", withConflicts, withoutConflicts)
	}
	if withConflicts.UnmergedCount != 1 || withoutConflicts.UnmergedCount != 0 {
		t.Fatalf("unmergedCount mismatch: %+v / %+v", withConflicts, withoutConflicts)
	}
}

func TestDescribeInProgress_Rebase(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{
		RebaseMergeDir: true, RebaseHeadName: strp("refs/heads/side"),
	}, nil)
	if got := gitpreflight.DescribeInProgress(op); got != "Rebasing side" {
		t.Fatalf("got %q", got)
	}
}

func TestDescribeInProgress_RevertWithSha(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{RevertHead: strp("d657c6ef00000000")}, nil)
	if got := gitpreflight.DescribeInProgress(op); got != "Reverting `d657c6e`" {
		t.Fatalf("got %q", got)
	}
}

func TestDescribeInProgress_UnmergedOnly(t *testing.T) {
	op := gitpreflight.ClassifyInProgress(gitpreflight.InProgressStateFiles{}, []string{"a.txt"})
	if got := gitpreflight.DescribeInProgress(op); got != "Unresolved conflict" {
		t.Fatalf("got %q", got)
	}
}

File: packages/git-core/src/model/operation.test.ts
import { describe, expect, test } from 'bun:test';
import {
  canRunOp,
  classifyInProgress,
  describeInProgress,
  type InProgressOperation,
  type InProgressStateFiles,
} from './operation.ts';

function stateFiles(partial: Partial<InProgressStateFiles>): InProgressStateFiles {
  return {
    mergeHead: undefined,
    cherryPickHead: undefined,
    revertHead: undefined,
    bisectLog: false,
    rebaseMergeDir: false,
    rebaseApplyDir: false,
    rebaseApplyApplying: false,
    rebaseHeadName: undefined,
    rebaseOnto: undefined,
    sequencerDir: false,
    sequencerTodoKind: undefined,
    ...partial,
  };
}

describe('classifyInProgress — precedence table (§7.11)', () => {
  test('no state files, no unmerged paths ⇒ null (nothing in progress)', () => {
    expect(classifyInProgress({ stateFiles: stateFiles({}), unmergedPaths: [] })).toBeNull();
  });

  // G26 D12 inverts this test (renamed from its G5 original, which asserted "cannot continue" —
  // that "§9 report-only posture" held only while nothing in the app could START a rebase;
  // stack.restack now does).
  test('rebase-merge dir present ⇒ rebase, can continue, can abort, can skip (G26 D12)', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({
        rebaseMergeDir: true,
        rebaseHeadName: 'refs/heads/side',
        rebaseOnto: 'deadbeef',
      }),
      unmergedPaths: ['a.txt'],
    });
    expect(op).toEqual({
      kind: 'rebase',
      otherSha: 'deadbeef',
      headName: 'refs/heads/side',
      conflictedPaths: ['a.txt'],
      canContinue: true,
      canAbort: true,
      isSequence: false,
      unmergedCount: 1,
      canSkip: true,
    });
  });

  test('rebase-apply dir present ⇒ rebase (am-based rebase), can continue (G26 D12)', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({ rebaseApplyDir: true }),
      unmergedPaths: [],
    });
    expect(op?.kind).toBe('rebase');
    expect(op?.canContinue).toBe(true);
  });

  // F3: a `git am` in progress shares rebase-apply/ with an apply-backend rebase, distinguished
  // only by the "applying" marker file -- without this, every Can* flag was wrongly true and
  // every offered action (`rebase --continue`/`--abort`/`--skip`) fails outright.
  test('rebase-apply dir with "applying" marker ⇒ am, report-only (F3)', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({
        rebaseApplyDir: true,
        rebaseApplyApplying: true,
        rebaseHeadName: 'refs/heads/side',
        rebaseOnto: 'deadbeef',
      }),
      unmergedPaths: ['a.txt'],
    });
    expect(op).toEqual({
      kind: 'rebase',
      otherSha: 'deadbeef',
      headName: 'refs/heads/side',
      conflictedPaths: ['a.txt'],
      canContinue: false,
      canAbort: false,
      isSequence: false,
      unmergedCount: 1,
      canSkip: false,
    });
  });

  test('MERGE_HEAD present ⇒ merge, can continue and abort', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({ mergeHead: 'cafebabe' }),
      unmergedPaths: ['b.txt'],
    });
    expect(op).toEqual({
      kind: 'merge',
      otherSha: 'cafebabe',
      headName: undefined,
      conflictedPaths: ['b.txt'],
      canContinue: true,
      canAbort: true,
      isSequence: false,
      unmergedCount: 1,
      canSkip: false,
    });
  });

  test('CHERRY_PICK_HEAD present ⇒ cherryPick', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({ cherryPickHead: 'c0ffee' }),
      unmergedPaths: [],
    });
    expect(op).toEqual({
      kind: 'cherryPick',
      otherSha: 'c0ffee',
      headName: undefined,
      conflictedPaths: [],
      canContinue: true,
      canAbort: true,
      isSequence: false,
      unmergedCount: 0,
      canSkip: true,
    });
  });

  test('REVERT_HEAD present ⇒ revert', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({ revertHead: 'f00dcafe' }),
      unmergedPaths: [],
    });
    expect(op).toEqual({
      kind: 'revert',
      otherSha: 'f00dcafe',
      headName: undefined,
      conflictedPaths: [],
      canContinue: true,
      canAbort: true,
      isSequence: false,
      unmergedCount: 0,
      canSkip: true,
    });
  });

  test('BISECT_LOG present ⇒ bisect, cannot continue, can abort', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({ bisectLog: true }),
      unmergedPaths: [],
    });
    expect(op).toEqual({
      kind: 'bisect',
      otherSha: undefined,
      headName: undefined,
      conflictedPaths: [],
      canContinue: false,
      canAbort: true,
      isSequence: false,
      unmergedCount: 0,
      canSkip: false,
    });
  });

  // F2: a paused multi-commit revert/cherry-pick whose current commit was finished with a plain
  // `git commit` (instead of `--continue`) clears *_HEAD but leaves sequencer/ (and its todo)
  // behind -- git's own state machine is still mid-run, so this must not fall through to
  // unmergedOnly (or null, with no unmerged paths left either).
  test('sequencer dir with no *_HEAD file, todo says revert ⇒ revert, not unmergedOnly (F2)', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({ sequencerDir: true, sequencerTodoKind: 'revert' }),
      unmergedPaths: [],
    });
    expect(op).toEqual({
      kind: 'revert',
      otherSha: undefined,
      headName: undefined,
      conflictedPaths: [],
      canContinue: true,
      canAbort: true,
      isSequence: true,
      unmergedCount: 0,
      canSkip: true,
    });
  });

  test('sequencer dir with no *_HEAD file, todo says pick ⇒ cherryPick (F2)', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({ sequencerDir: true, sequencerTodoKind: 'cherryPick' }),
      unmergedPaths: [],
    });
    expect(op?.kind).toBe('cherryPick');
    expect(op?.isSequence).toBe(true);
  });

  test('sequencer dir alone, todo unreadable/unrecognised ⇒ falls through to null (unchanged)', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({ sequencerDir: true, sequencerTodoKind: undefined }),
      unmergedPaths: [],
    });
    expect(op).toBeNull();
  });

  test('unmerged paths with none of the six state files ⇒ unmergedOnly, cannot continue or abort', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({}),
      unmergedPaths: ['c.txt', 'd.txt'],
    });
    expect(op).toEqual({
      kind: 'unmergedOnly',
      otherSha: undefined,
      headName: undefined,
      conflictedPaths: ['c.txt', 'd.txt'],
      canContinue: false,
      canAbort: false,
      isSequence: false,
      unmergedCount: 2,
      canSkip: false,
    });
  });

  test('shadowing: rebase state files present alongside cherry-pick-shaped sequencer state ⇒ still rebase', () => {
    // A rebase --onto stopped on a conflict leaves rebase-merge/ present; if the sequencer dir
    // and even a stray CHERRY_PICK_HEAD-like signal were also present, rebase must still win —
    // it is checked first in the precedence table, unconditionally.
    const op = classifyInProgress({
      stateFiles: stateFiles({
        rebaseMergeDir: true,
        rebaseHeadName: 'refs/heads/topic',
        rebaseOnto: 'abc123',
        sequencerDir: true,
        cherryPickHead: 'shouldnotwin',
      }),
      unmergedPaths: [],
    });
    expect(op?.kind).toBe('rebase');
    expect(op?.isSequence).toBe(true);
  });

  test('shadowing: merge/cherryPick/revert all present at once ⇒ merge wins (table order)', () => {
    const op = classifyInProgress({
      stateFiles: stateFiles({
        mergeHead: 'merge-sha',
        cherryPickHead: 'cherry-sha',
        revertHead: 'revert-sha',
      }),
      unmergedPaths: [],
    });
    expect(op?.kind).toBe('merge');
  });

  test('fallthrough: unmergedOnly only applies when none of the five state-file kinds matched', () => {
    // Bare unmerged paths with a bisect log present must classify as bisect, not unmergedOnly —
    // bisect is checked before the fallback.
    const op = classifyInProgress({
      stateFiles: stateFiles({ bisectLog: true }),
      unmergedPaths: ['e.txt'],
    });
    expect(op?.kind).toBe('bisect');
    expect(op?.unmergedCount).toBe(1);
  });
});

describe('canRunOp — the gate (§7.11)', () => {
  test('nothing in progress: every op kind is allowed', () => {
    expect(canRunOp(null, 'checkout')).toBe(true);
    expect(canRunOp(null, 'revert')).toBe(true);
    expect(canRunOp(null, 'branchCreate')).toBe(true);
    expect(canRunOp(null, 'tagDelete')).toBe(true);
  });

  const inProgress: InProgressOperation = {
    kind: 'merge',
    otherSha: 'sha',
    headName: undefined,
    conflictedPaths: ['a'],
    canContinue: true,
    canAbort: true,
    isSequence: false,
    unmergedCount: 1,
    canSkip: false,
  };

  test('in progress: checkout is gated', () => {
    expect(canRunOp(inProgress, 'checkout')).toBe(false);
  });

  test('in progress: revert is gated', () => {
    expect(canRunOp(inProgress, 'revert')).toBe(false);
  });

  test('in progress: branch/tag creation and deletion are NOT gated (git allows them)', () => {
    expect(canRunOp(inProgress, 'branchCreate')).toBe(true);
    expect(canRunOp(inProgress, 'branchDelete')).toBe(true);
    expect(canRunOp(inProgress, 'branchRename')).toBe(true);
    expect(canRunOp(inProgress, 'tagCreate')).toBe(true);
    expect(canRunOp(inProgress, 'tagDelete')).toBe(true);
    expect(canRunOp(inProgress, 'tagPush')).toBe(true);
    expect(canRunOp(inProgress, 'tagDeleteRemote')).toBe(true);
  });

  test('in progress: opContinue/opAbort are not gated', () => {
    expect(canRunOp(inProgress, 'opContinue')).toBe(true);
    expect(canRunOp(inProgress, 'opAbort')).toBe(true);
  });

  // P10 probe 3: git does not refuse `reset --mixed`/`--hard` mid-merge on its own — it silently
  // abandons the sequencer state — so `reset` must be gated here even though no earlier phase's
  // op needed host-side enforcement to match git's own refusal.
  test('in progress: reset and cherryPick are gated (probe 3)', () => {
    expect(canRunOp(inProgress, 'reset')).toBe(false);
    expect(canRunOp(inProgress, 'cherryPick')).toBe(false);
  });

  // `opSkip` is only ever offered *from within* the gated state itself (`canSkip`), never a way
  // to bypass it — so it is deliberately NOT in GATED_OP_KINDS.
  test('in progress: opSkip is not gated', () => {
    expect(canRunOp(inProgress, 'opSkip')).toBe(true);
  });
});

describe('classifyInProgress — canSkip (P10 probe 6, G26 D12)', () => {
  test('cherryPick, revert and rebase can skip; every other kind cannot', () => {
    expect(
      classifyInProgress({
        stateFiles: stateFiles({ cherryPickHead: 'c0ffee' }),
        unmergedPaths: [],
      })?.canSkip,
    ).toBe(true);
    expect(
      classifyInProgress({ stateFiles: stateFiles({ revertHead: 'f00d' }), unmergedPaths: [] })
        ?.canSkip,
    ).toBe(true);
    expect(
      classifyInProgress({ stateFiles: stateFiles({ mergeHead: 'm' }), unmergedPaths: [] })
        ?.canSkip,
    ).toBe(false);
    // G26 D12 inverts this line: rebase now offers Skip too (git's own conflict hint names
    // `git rebase --skip` verbatim).
    expect(
      classifyInProgress({
        stateFiles: stateFiles({ rebaseMergeDir: true }),
        unmergedPaths: [],
      })?.canSkip,
    ).toBe(true);
    expect(
      classifyInProgress({ stateFiles: stateFiles({ bisectLog: true }), unmergedPaths: [] })
        ?.canSkip,
    ).toBe(false);
    expect(classifyInProgress({ stateFiles: stateFiles({}), unmergedPaths: ['x'] })?.canSkip).toBe(
      false,
    );
  });
});

describe('describeInProgress', () => {
  function op(
    partial: Partial<InProgressOperation> & Pick<InProgressOperation, 'kind'>,
  ): InProgressOperation {
    return {
      otherSha: undefined,
      headName: undefined,
      conflictedPaths: [],
      canContinue: false,
      canAbort: true,
      isSequence: false,
      unmergedCount: 0,
      canSkip: false,
      ...partial,
    };
  }

  test('merge: label plus truncated sha', () => {
    expect(describeInProgress(op({ kind: 'merge', otherSha: '0123456789abcdef' }))).toBe(
      'Merging `0123456`',
    );
  });

  test('cherryPick: label plus truncated sha', () => {
    expect(describeInProgress(op({ kind: 'cherryPick', otherSha: 'fedcba9876543210' }))).toBe(
      'Cherry-picking `fedcba9`',
    );
  });

  test('revert: label plus truncated sha', () => {
    expect(describeInProgress(op({ kind: 'revert', otherSha: 'aaaaaaaaaaaa' }))).toBe(
      'Reverting `aaaaaaa`',
    );
  });

  test('bisect: label, no sha to show', () => {
    expect(describeInProgress(op({ kind: 'bisect' }))).toBe('Bisecting');
  });

  test('rebase: strips the refs/heads/ prefix from headName', () => {
    expect(
      describeInProgress(op({ kind: 'rebase', headName: 'refs/heads/feature', otherSha: 'sha' })),
    ).toBe('Rebasing feature');
  });

  test("rebase: falls back to bare 'Rebasing' when headName is undefined", () => {
    expect(describeInProgress(op({ kind: 'rebase', headName: undefined }))).toBe('Rebasing');
  });

  test('unmergedOnly: fixed sentence, ignores any otherSha', () => {
    expect(describeInProgress(op({ kind: 'unmergedOnly' }))).toBe('Unresolved conflict');
  });

  test('a kind with no otherSha omits the trailing sha entirely (no dangling backticks)', () => {
    expect(describeInProgress(op({ kind: 'merge', otherSha: undefined }))).toBe('Merging');
  });
});
